packages/vite/src/node/config.ts

```diff
@@ -603,7 +604,11 @@ export function resolveDevEnvironmentOptions(
       preserverSymlinks,
       environmentName,
     ),
-    createEnvironment: dev?.createEnvironment,
+    createEnvironment:
+      dev?.createEnvironment ??
+      (environmentName === 'ssr'
+        ? createNodeSsrDevEnvironment
+        : createDevEnvironment),
```

Comment:
`ssr` is a special backward compat environment, but reverts to a regular environment if `createEnvironment` is defined. If not, by default it uses `createNodeSsrDevEnvironment`, that inside uses `hot: createServerHotChannel()` and then goes to `createNodeDevEnvironment()`.